backend/zoho/zoho.go

```diff
@@ -0,0 +1,1223 @@
+func (f *Fs) Put(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (fs.Object, error) {
+	existingObj, err := f.newObjectWithInfo(ctx, src.Remote(), nil)
+	switch err {
+	case nil:
+		return existingObj, existingObj.Update(ctx, in, src, options...)
+	case fs.ErrorObjectNotFound:
+		// Not found so create it
+		return f.PutUnchecked(ctx, in, src)
+	default:
+		return nil, err
+	}
+}
+
+func isSimpleName(s string) bool {
+	for _, r := range s {
+		if (r < 'a' || r > 'z') && (r < 'A' || r > 'Z') && (r != '.') {
+			return false
+		}
+	}
+	return true
+}
+
+func (f *Fs) upload(ctx context.Context, name string, parent string, size int64, in io.Reader, options ...fs.OpenOption) (*api.Item, error) {
+	params := url.Values{}
+	params.Set("filename", name)
+	params.Set("parent_id", parent)
+	params.Set("override-name-exist", strconv.FormatBool(true))
+	formReader, contentType, overhead, err := rest.MultipartUpload(in, nil, "content", name)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to make multipart upload")
+	}
+
+	contentLength := overhead + size
+	opts := rest.Opts{
+		Method:           "POST",
+		Path:             "/upload",
+		Body:             formReader,
+		ContentType:      contentType,
+		ContentLength:    &contentLength,
+		Options:          options,
+		Parameters:       params,
+		TransferEncoding: []string{"identity"},
+	}
+
+	var resp *http.Response
+	var uploadResponse *api.UploadResponse
+	err = f.pacer.CallNoRetry(func() (bool, error) {
+		resp, err = f.srv.CallJSON(ctx, &opts, nil, &uploadResponse)
+		return shouldRetry(resp, err)
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "upload error")
+	}
+	if len(uploadResponse.Uploads) != 1 {
+		return nil, errors.New("upload: invalid response")
+	}
+	// Received meta data is missing size so we have to read it again.
+	info, err := f.readMetaDataForID(ctx, uploadResponse.Uploads[0].Attributes.RessourceID)
+	if err != nil {
+		return nil, err
+	}
+
+	return info, nil
+}
+
+// PutUnchecked the object into the container
+//
+// This will produce an error if the object already exists
+//
+// Copy the reader in to the new object which is returned
+//
+// The new object may have been created if an error is returned
+func (f *Fs) PutUnchecked(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (fs.Object, error) {
+	size := src.Size()
+	remote := src.Remote()
+
+	// Create the directory for the object if it doesn't exist
+	leaf, directoryID, err := f.dirCache.FindPath(ctx, remote, true)
+	if err != nil {
+		return nil, err
+	}
+
+	if isSimpleName(leaf) {
+		info, err := f.upload(ctx, f.opt.Enc.FromStandardName(leaf), directoryID, size, in, options...)
+		if err != nil {
+			return nil, err
+		}
+		return f.newObjectWithInfo(ctx, remote, info)
+	}
+
+	tempName := "rcloneTemp" + random.String(8)
+	info, err := f.upload(ctx, tempName, directoryID, size, in, options...)
```

Comment:
Zoho's upload endpoint has problems with special characters even though they are supported by every other API endpoint. For uploads, the filename is supplied both as a multipart param and an URL parameter. The multipart filename parameter cannot handle `/`, `\`, `"` and `'` and no way of escaping these characters seems to work. The URL parameter also has problems with an even larger set of special characters even if they are-URL encoded. 
This forced me to implement a workaround that uploads files with a temporary name and rename them afterward. 
The `isSimpleName` path is a shortcut without rename used for files that only contain alphanumeric characters and `.` which should be the majority of filenames anyway.

Reply:
Grrr, that is annoying! Your workaround seems like an efficient one.

Reply:
Really? That is strange. Maybe it it is supporting only a latin1 subset wrapped in UTF-8 or something like that?

Reply:
It's weirder than that cyrillic and chinese are supported.